[EuiResizableContainer] Allow collapsible panels

andreadelrio opened the pull request:
### Summary

Adds the ability for `EuiResizablePanel` components to be collapsible—shrinking the panel itself and hiding its content either via an out-of-the-box toggle button or through coordinated actions hooks provided to the parent component.

<img width="946" alt="image" src="https://user-images.githubusercontent.com/2728212/98745020-f8f3c780-236f-11eb-984a-34660b3e7bd1.png">

**Features:**
* New `mode` prop specifies an `EuiResizablePanel` as `collapsible` or `main`, indicating the ability to collapse or the preference to gain space from collapsed panels.
* Incorporated `EuiPanel` as the base element, allowing for more panel style options
* Rewrote registry and interaction logic to use a React reducer pattern, making evening more reliable and providing more flexibility in exposing action hooks to consumers.

**Breaking change:**
Removed the `size` prop on `EuiResizableButton`. This was originally added as something of a spacing mechanism, but is less effective than the newly added `panelProp.paddingSize` prop.


### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**
- [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples
- [x] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**
- [x] Checked for **breaking changes** and labeled appropriately
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

cchaos commented:
🎉 Great start! I think this will definitely meet Discover's needs and replicate a similar behavior that exists now. 

I'd love to see a version that allows the toggle button to be free-floating, meaning, it doesn't have to live in the DOM layout ([possibly pushing content down](https://share.getcloudapp.com/6quebRAm)).

A somewhat static version could be just adding the button on focus that floats in the center where the original grab button was. For instance, we'd have these 4 states (the fourth showing the collapsed state that maybe **always** shows the uncollapse button).

<img width="903" alt="Screen Shot 2020-08-31 at 17 02 38 PM" src="https://user-images.githubusercontent.com/549577/91768589-d42ee700-ebab-11ea-8014-f7731520c543.png">

^^ That solution will work well with applications that are ok keeping at least a 24px wide area dedicated to the collapsible button.

However, we'd probably also need a solution for consumers who would like to provide their own toggle button **anywhere** in their UI that would then completely hide or show this panel (similar to like a flyout toggle). You could handle this by allowing `toggle` to accept any ReactElement which will then add the `onClick` behavior to the button they pass in.

One more enhancement could be to add a double-click behavior which will do the collapsing/uncollapsing behavior so that users don't have to hunt for the button itself. It would be a **learned** behavior, but that is ok since we're also providing the targetable button.

chandlerprall commented:
Because the collapse functionality - defined on one or more panels - affects the other panel(s) in a group, I think the majority of the logic needs to move to the panel registry which is already responsible for orchestrating the group.

Some additional concerns:

* providing collapse/uncollapse feedback to the application
* allowing the app to provide initial collapsed state and/or complete control during the panel lifespan, like we do for widths
* when collapsed, what happens to reported widths? (see below)

#### Reported widths

As an example let's start with 3 panels all open and with 33% width.

```
|   A   |   B   |   C   |
|       |       |       |
|       |       |       |
```

Let's collapse the middle column. The total displayed width should remain the same.

```
|     A    |*|     C    |
|          | |          |
|          | |          |
```

Now resize column C larger

```
|   A   |*|      C      |
|       | |             |
|       | |             |
```

Numerically, what happened?

Today, with un-collapsible columns, resizing C would take away from B. Collapsed, we cannot take away from B because visually it wouldn't adjust C's width, because B is already collapsed. Does it look over further and take away from A? What sizes are expected when we expand B again? Would any of those answers be meaningfully different if we had shrunk C instead of enlarging?

andreadelrio commented:
> Because the collapse functionality - defined on one or more panels - affects the other panel(s) in a group, I think the majority of the logic needs to move to the panel registry which is already responsible for orchestrating the group.
> 
> Some additional concerns:
> 
> * providing collapse/uncollapse feedback to the application
> * allowing the app to provide initial collapsed state and/or complete control during the panel lifespan, like we do for widths
> * when collapsed, what happens to reported widths? (see below)
> 
> #### Reported widths
> As an example let's start with 3 panels all open and with 33% width.
> 
> ```
> |   A   |   B   |   C   |
> |       |       |       |
> |       |       |       |
> ```
> 
> Let's collapse the middle column. The total displayed width should remain the same even though A and C together still represent only 66% of the size.
> 
> ```
> |     A    |*|     C    |
> |          | |          |
> |          | |          |
> ```
> 
> Now resize column C larger
> 
> ```
> |   A   |*|      C      |
> |       | |             |
> |       | |             |
> ```
> 
> Numerically, what happened?
> 
> Today, with un-collapsible columns, resizing C would take away from B. Collapsed, we cannot take away from B because visually it wouldn't adjust C's width, because B is already collapsed (this is not strictly true, as proportionally A would still shrink to accommodate C's increased size, but that feels really awkward from a user's perspective). Does it look over further and take away from A? What sizes are expected when we expand B again? Would any of those answers be meaningfully different if we had shrunk C instead of enlarging?

I made a diagram to better understand the situation

<img width="953" alt="Group 3" src="https://user-images.githubusercontent.com/4016496/92019125-f68b4680-ed0a-11ea-97be-1ab98b22fc9e.png">

I agree that changing the behavior of the `EuiResizableButton` upon collapsing one panel could be confusing but what would be the alternative? The other question that came to mind, should there we "limits" as to when to implement collapsible panels? What if consumers add `toggle` to two panels or to three? I think that we could fall into a loop of constantly changing the behavior of the `EuiResizableButton`s and that could get messy.

chandlerprall commented:
> I agree that changing the behavior of the EuiResizableButton upon collapsing one panel could be confusing but what would be the alternative?

Not saying we _shouldn't_ change the behaviour, only that we need to decide on what that change should be. I did a quick search to find this pattern elsewhere and didn't find anything. Perhaps the solution, at least for now, is disable the resize button if it borders a collapsed panel?

andreadelrio commented:
> > I agree that changing the behavior of the EuiResizableButton upon collapsing one panel could be confusing but what would be the alternative?
> 
> Not saying we _shouldn't_ change the behaviour, only that we need to decide on what that change should be. I did a quick search to find this pattern elsewhere and didn't find anything. Perhaps the solution, at least for now, is disable the resize button if it borders a collapsed panel?

@chandlerprall I think that could be a good option. In particular, I think that when there are only two panels it wouldn't feel weird to the user. That's the use case in Discover. Things get trickier when there's more than two panels. 

I found a similar interface [elsewhere](https://split.js.org
). There, because the UI is a bit different (basically the "resizable button" is the panels’ borders) it’s more clear that it makes sense to disable resizing of a panel if it borders one that's been collapsed.

In the gif below, if I could collapse HTML I wouldn’t be able to further resize JS or CSS and that would feel ok. JS only affects HTML (and viceversa) and CSS only affects JS (and viceversa). With HTML collapsed, nothing should affect the JS panel's width. 

![beh](https://user-images.githubusercontent.com/4016496/92163705-fc575980-ede8-11ea-86cc-04aff8b2f7a8.gif)

chandlerprall commented:
Works for me - let's go the route of disabling resizing any panel that neighbors a collapsed one. We'll get you engineering support on the panel registry changes.

thompsongl commented:
> Works for me - let's go the route of disabling resizing any panel that neighbors a collapsed one. We'll get you engineering support on the panel registry changes.

I'll most likely be the one helping here. A question that immediately comes to mind:
If we move the logic up to the panel registry, then it seems like `willExpand` can be removed entirely. Unless there are situations where a panel next to a `toggle` panel should not expand, but that seems unlikely (what would happen if the neighboring panel _didn't_ expand?). Does `willExpand` always need to accompany `toggle`?

chandlerprall commented:
> it seems like willExpand can be removed entirely

Good catch, that's my interpretation as well.

andreadelrio commented:
Awesome. I think ideally `willExpand` should not exist. I only added it because it was the only way I found to mark which panels would be giving up their width but I was working under the assumption that all _accompanying panels would be giving up their width_. So `toggle` alone should be enough.

> I'll most likely be the one helping here. A question that immediately comes to mind:
> If we move the logic up to the panel registry, then it seems like `willExpand` can be removed entirely. Unless there are situations where a panel next to a `toggle` panel should not expand, but that seems unlikely (what would happen if the neighboring panel _didn't_ expand?). Does `willExpand` always need to accompany `toggle`?

thompsongl commented:
> I think ideally `willExpand` should not exist. [...] So toggle alone should be enough.

Perfect

thompsongl commented:
Question on sizing for sibling panels after a panel is collapsed (with 3 or more panels):

Given:
```
|   A   |   B   |   C   |
|       |       |       |
|       |       |       |
```

And B collapses, above examples show as below, where A and C grow equally:
```
|     A    |*|     C    |
|          | |          |
|          | |          |
```

But what about collapsing A or C instead? Do we expect all sibling panels to grow equally even if not a direct sibling?
```
|*|     B    |     C    |
| |          |          |
| |          |          |
```

or
```
|*|       B      |    C   |
| |              |        |
| |              |        |
```

If all siblings do not grow equally, does that change the assumption about A and C growing equally in the first example?

thompsongl commented:
This ^ all gets even more tricky if we try equal panel growth, but not all panels were sized the same to start.

cchaos commented:
I like to try to think these through in terms of real-world scenario/use-cases. For instance, I have a need to use these in Lens to be able to allow the user to resize the left and right sidebars as well as fully collapse them if they need the space. The sidebars have defined widths and the center takes up the remaining.

<img width="50%" alt="Screen Shot 2020-09-10 at 18 28 24 PM" src="https://user-images.githubusercontent.com/549577/92816596-b6276a80-f393-11ea-83bf-1a4ebd9e11e0.png">

When the user adjust the right side panel, it is resizing the right sidebar and the center is compensating.

<img width="50%" alt="Screen Shot 2020-09-10 at 18 32 44 PM" src="https://user-images.githubusercontent.com/549577/92817051-2b933b00-f394-11ea-8e29-4bf074a2cf7e.png">

When the user collapses the right sidebar, the same thing happens. It adjusts the sizing of the right sidebar while the center grows to fill the space.

<img width="50%" alt="Screen Shot 2020-09-10 at 18 28 11 PM" src="https://user-images.githubusercontent.com/549577/92817200-51204480-f394-11ea-865d-bbf02a37f7bc.png">

Here is a [video](https://share.getcloudapp.com/8LuPKBOv) and a [prototype](https://www.figma.com/proto/CJfejj3KoCnJqx58wzfCvf/Flows?node-id=895%3A426&viewport=25%2C424%2C0.4534960389137268&scaling=scale-down) of the collapsing behavior (drag to initiate toggle).

---

Essentially when one panel adjusts only it's direct sibling  adjusts as well. If it has 2 siblings, some maths to find the percentage of growth needed to add to each panel?

thompsongl commented:
> If it has 2 siblings, some maths to find the percentage of growth needed to add to each panel?

Yeah, it's the 2+ sibling case that I was really asking about. I think adding to each sibling makes sense, just want to be sure

thompsongl commented:
Spent some time trying outlining a spec for behavior this morning, and I have some more questions 🙂 

Based on all the discussion in this thread, it seems there are 2 ways to implement a toggle button, each having different impacts on expected behavior.


Use the following for both:
```
|   A   |   B   |   C   |
|       |       |       |
|       |       |       |
```
___

**First: Toggle button on the drag handle**

Given the example where B and C both have `toggle=true`, both drag handles would have toggle buttons:

Directionality is therefore implied.

* The handle between A and B would have a single button
  * It would point to the right, indicating that B would collapse on click
* The handle between B and C would have two buttons
  * One pointing left, indicating that B would collapse on click
  * One pointing right, indicating that C would collapse on click
* Width would only be redistributed to the panel bordering the drag handle
  * For example, when collapsing C, only B would grow
  * When collapsing B, it growth would depend on which drag handle button was clicked

___

**Second: Toggle button anywhere except the drag handle**

Given the example where B and C both have `toggle=true`, and we place a single toggle button for each panel somewhere else in the UI. For example, "Collapse/Expand panel B" and that button is clicked:

Directionality cannot be assumed like in the first example.

* Width could be redistributed in various ways
  * We could choose A, or C, or both to grow
  * We could provide additional configuration options here
* Depending on which choices are made, the default toggle buttons inside the panels need to adapt


___

* Are both of these options desired or am I off base?
* Do the details that I've described for each sound correct?
* If both, should both be possible at once (drag handle buttons and buttons elsewhere)?

Regardless of above, I assume we also want to restore the previous width of a collapsed panel after it again becomes expanded?

andreadelrio commented:
> Spent some time trying outlining a spec for behavior this morning, and I have some more questions 🙂
> 
> Based on all the discussion in this thread, it seems there are 2 ways to implement a toggle button, each having different impacts on expected behavior.
> 
> Use the following for both:
> 
> ```
> |   A   |   B   |   C   |
> |       |       |       |
> |       |       |       |
> ```
> 
> **First: Toggle button on the drag handle**
> 
> Given the example where B and C both have `toggle=true`, both drag handles would have toggle buttons:
> 
> Directionality is therefore implied.
> 
> * The handle between A and B would have a single button
>   
>   * It would point to the right, indicating that B would collapse on click
> * The handle between B and C would have two buttons
>   
>   * One pointing left, indicating that B would collapse on click
>   * One pointing right, indicating that C would collapse on click
> * Width would only be redistributed to the panel bordering the drag handle
>   
>   * For example, when collapsing C, only B would grow
>   * When collapsing B, it growth would depend on which drag handle button was clicked
> 
> **Second: Toggle button anywhere except the drag handle**
> 
> Given the example where B and C both have `toggle=true`, and we place a single toggle button for each panel somewhere else in the UI. For example, "Collapse/Expand panel B" and that button is clicked:
> 
> Directionality cannot be assumed like in the first example.
> 
> * Width could be redistributed in various ways
>   
>   * We could choose A, or C, or both to grow
>   * We could provide additional configuration options here
> * Depending on which choices are made, the default toggle buttons inside the panels need to adapt
> 
> * Are both of these options desired or am I off base?
> * Do the details that I've described for each sound correct?
> * If both, should both be possible at once (drag handle buttons and buttons elsewhere)?
> 

@thompsongl I personally don't have a strong preference for either implementation. I'd say go with the simplest one which it sounds like it's the first one?

> Regardless of above:
> 
> * Do we want the ability to collapse multiple panels at once?

I think definitely. Think of the Lens use case Caroline shared above, the sidebar should also be collapsible.

> * I assume we also want to restore the previous width of a collapsed panel after it again becomes expanded?

Ideally yes.

thompsongl commented:
Summary of sync discussion:

**More configuration**
* We want consumers to have fine-grained control
  * Button appearance and position
  * Panel size redistribution
* But also need to have sensible defaults/zero-config ability
  * Buttons hover over drag handle by default
* Greg will explore what an API will look like

**More flexibility**
* Multiple panels can be collapsed simultaneously
* External control for toggling behavior, perhaps even full external control

**More context structure**
* The registry system will need to account for the relationships between panels and drag handles, which it currently does not
* Need to store previous state in order to restore panel sizing

While breaking changes can likely be avoided, there will need to be substantial changes made to the registry and context system.

thompsongl commented:
_Update:_

Still in progress. Current status below, but wanted to give @chandlerprall an early look at the refactor before going even further.

**Major refactor**
* Transformed 'helper' pattern into a reducer pattern via `useReducer` & custom hook
* Changed helper/reducer actions to be less dependent on DOM events
* Exposed reducer actions to consumer via render prop (see _To do_ section)
* Moved class-based registry into the reducer state

**Collapsibles**
* Converted from CSS-based to JS-based via reducer pattern
* Introduced more state into panel and resizer references

**To do**
* Establish `toggle` API for consumer configuration
* Establish better action names
* Expose a clearer set of actions to consumers
* Write more  documentation
* Write more tests

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

chandlerprall commented on src/components/resizable_container/resizable_button.tsx:
```diff
@@ -89,17 +105,20 @@ export const EuiResizableButton: FunctionComponent<EuiResizableButtonProps> = ({
   const previousRef = useRef<HTMLElement>();
   const onRef = useCallback(
     (ref: HTMLElement | null) => {
+      if (!register || !deregister) return;
```
Maybe make these two functions part of a single incoming prop configuration so someone doesn't add one but not the other?

```
<EuiResizableButton registerCbs={{ ... }}/>
```

Or change the TS interface to enforce?

thompsongl replied:
Good call. I'll look to one of these options.

chandlerprall reviewed (commented):
Overall structure is 👍 The `todo` questions sound good and I look forward to those answers 😄

thompsongl commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -238,5 +241,42 @@ export const ResizableContainerExample = {
+        <div>
+          <p>
+            You can add toggling to <strong>EuiResizablePanel</strong>s inside a{' '}
+            <strong>EuiResizableContainer</strong> of horizontal{' '}
```
Ah yep. FWIW, the docs are out of date, but I do recall in code that collapsing is restricted to horizontal orientations. I'll update the todo list

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

thompsongl commented:
@cchaos @andreadelrio I pushed some WIP changes that show concepts for button placement and directional panel collapsing. Simpler cases (2-3 panels) work well and I think it's pretty clear to users how collapsing will work. More complex cases (4+ panels) still work, but I think it's much less clear what to expect. Looking for design help on making actions more clear, including possibly placing limits on what can/cannot be collapsed.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

andreadelrio commented:
> @cchaos @andreadelrio I pushed some WIP changes that show concepts for button placement and directional panel collapsing. Simpler cases (2-3 panels) work well and I think it's pretty clear to users how collapsing will work. More complex cases (4+ panels) still work, but I think it's much less clear what to expect. Looking for design help on making actions more clear, including possibly placing limits on what can/cannot be collapsed.

From a design perspective I think it'd be nice to have two options:

a) position: determines the placement of the toggle button. Options: `top`, `center` and `bottom`. When the position is `center`, the toggle button (visually) replaces the drag icon.
b) showOnFocus: boolean. If it’s `true` the toggle button is only visible when the `EuiResizableButton`
has focus (which is one option @cchaos suggested earlier).

A case with `position="center"` and `showOnFocus={false}` would then look and behave like this:
![Screen Recording 2020-10-05 at 4 06 06 PM](https://user-images.githubusercontent.com/4016496/95146057-a9e3c280-0731-11eb-95e2-43e6c107e621.gif)


Also, after looking at the example of 4 panels where all have `toggle`, I’m of the mind of only showing 1 toggle button per panel. I guess we’d arbitrarily pick which toggle button to keep?

thompsongl commented:
> I’m of the mind of only showing 1 toggle button per panel. I guess we’d arbitrarily pick which toggle button to keep?

We'd probably need to introduce a `direction` option in this case. The problem I see with an API per panel like `{ position, direction } is that overlapping `position` could cause  overlapping buttons, and `direction="right"` is never valid for the first panel (just as `direction="left"` is never valid for the last). So it'd be really easy for consumers to misconfigure collapsible options. It seems to me that we need a default/zero-config case that accounts for all panels being collapsible that we're happy with from a design perspective, or we have no default cause and require the consumer to make all choices.

A different approach altogether is to not provide collapsible buttons at all. We now expose a collapse-panel-by-ID function that consumers could use to implement their own actions, at which point we could document guidance but not force a UI. 

Another point, we should never allow all panels to be collapsed at once. Otherwise we'll need to consider some kind of empty state.

andreadelrio commented:
> > I’m of the mind of only showing 1 toggle button per panel. I guess we’d arbitrarily pick which toggle button to keep?
> 
> We'd probably need to introduce a `direction` option in this case. The problem I see with an API per panel like `{ position, direction }` is that overlapping `position` could cause overlapping buttons, and `direction="right"` is never valid for the first panel (just as `direction="left"` is never valid for the last). So it'd be really easy for consumers to misconfigure collapsible options. It seems to me that we need a default/zero-config case that accounts for all panels being collapsible that we're happy with from a design perspective, or we have no default cause and require the consumer to make all choices.
> The `direction` option could move to a higher level and only be set once, considering the invalidity of some choices.
> 
> A different approach altogether is to not provide collapsible buttons at all. We now expose a collapse-panel-by-ID function that consumers could use to implement their own actions, at which point we could document guidance but not force a UI.
> 


@thompsongl 
- Can you clarify the scenario where overlapping buttons would occur?
- What if we only allow `direction` for middle panels? We'd pick a default `direction` but let the consumer choose a different direction for a middle panel if they need to.

> Another point, we should never allow all panels to be collapsed at once. Otherwise we'll need to consider some kind of empty state.

Maybe this is a good option, that way consumers can choose what they want the toggle element to be as well as control its position. Do you foresee any major downsides to this approach?

thompsongl commented:
Pushed another WIP commit that shows the `showOnFocus` functionality "working". Might be helpful to see a less cluttered UI when considering design.

> Can you clarify the scenario where overlapping buttons would occur?

If a consumer specifies `position: 'top'` for all panels, buttons would overlap each other (occupy the same space) when a resizer has two buttons. 

> What if we only allow direction for middle panels?

Certainly an option, but configuration just seems confusing when certain options work for one panel and not another.

> Maybe this is a good option, that way consumers can choose what they want the toggle element to be as well as control its position. Do you foresee any major downsides to this approach?

Lack of consistency in design across implementations. Also, it'd be difficult to achieve a show-on-focus button like in @cchaos mocks because of the depth of event handling that needs to happen for a11y.

andreadelrio commented:
> > Can you clarify the scenario where overlapping buttons would occur?
> 
> If a consumer specifies `position: 'top'` for all panels, buttons would overlap each other (occupy the same space) when a resizer has two buttons.
> 
> > What if we only allow direction for middle panels?
> 
> Certainly an option, but configuration just seems confusing when certain options work for one panel and not another.

I see, when I suggested `position` I was thinking each panel would have one toggle button tops. How important is it to add `direction` versus picking one for them for middle panels? Would making that decision on behalf of consumers be too prescriptive? 

The way I see it, a use case with multiple panels where the consumer sets `toggle` on all of them (meaning including "middle" panels) is a bit of an edge case. From the Kibana use cases I can think of for toggle, it is not usually set in all panels:

**Discover**

A (toggle) | B

**Visualize**

A | B (toggle)

**Lens**

A (toggle) | B | C (toggle)

My suggestion would be to only show one toggle per panel and not have `direction`. But might be worth seeing what Chandler or Caroline think.

> Lack of consistency in design across implementations. Also, it'd be difficult to achieve a show-on-focus button like in @cchaos mocks because of the depth of event handling that needs to happen for a11y.

That was my concert too (show-on-focus), sounds like this wouldn't be the way to go then.

thompsongl commented:
> The way I see it, a use case with multiple panels where the consumer sets toggle on all of them (meaning including "middle" panels) is a bit of an edge case

I think this is the fundamental question: do we intend this to function essentially like a big, layout-level accordion component? I think the answer is no.

Which brings up a new idea: what if we change the concept of the `toggle` prop? Instead introduce a `mode` (or something) prop that is `'collapsible' | 'main' | 'default'`. A `main` panel indicates that the panel is not collapsible and informs how the `collapsible` panels work. One `main` panel would be required if there are one or more `collapsible` panels in use. We'd likely be able to infer size redistribution, directionality, etc., without additional configuration. The Discover, Visualize, and Lens examples fit this idea well: the have one main content panel that is always relevant, and one or more action panels that are not always relevant.

We could tailor the default UI to this idea (see all single-toggle-button mocks in the thread), and expose methods for consumers to use if they wish to implement their own layout that falls outside of Discover-Visualize-Lens paradigm.

And, of course, the default is a `default` panel and this component works just as it does today.

cc @chandlerprall and @cchaos for a check on my thought process here

chandlerprall commented:
I like the `mode` approach, though I'm not sure what the _default_ specification really means? I feel _main_ would be a workable default?

> A main panel indicates that the panel is not collapsible and informs how the collapsible panels work

At first glance, I agree. A thought exercise: what if there are two `main`s?

```
A[collapsible] **|** B[main] **|** C[collapsible] **|** D[main] 
```

Visually, how does `A` or `B` collapse? I suppose there's a third option: `left|right|inward` ?

thompsongl commented:
> I feel main would be a workable default?

Good point here. Specifying `mode` in this way was my method of disallowing all panels being collapsible (i.e., validate the configuration on mount and/or with types), and directing collapsible panel direction (i.e., panels always collapse _away_ from the `main` panel).

> Visually, how does A or B collapse?

I assume you meant A or C (B would not be collapsible as a `main` panel). A would collapse to the left, and C would also probably collapse to the left. I think I'd like to not introduce a way for this to be configurable.

And perhaps we should only allow one `main`? That would simplify the answer to both questions.

chandlerprall commented:
> I assume you meant A or C (B would not be collapsible as a main panel). A would collapse to the left, and C would also probably collapse to the left. I think I'd like to not introduce a way for this to be configurable.
> 
> And perhaps we should only allow one main? That would simplify the answer to both questions.

Both of those sounds good to me.

thompsongl commented:
> Both of those sounds good to me, to limit other decisions & scope creep.

Pushed changes that show the logic for a `mode` prop and single `main` panel. The code is _not great_ right now, but you get a much better idea of design.

andreadelrio commented:
@thompsongl I'm liking the way `mode` behaves. Two comments: 

![image 525](https://user-images.githubusercontent.com/4016496/96077421-9d174b00-0e64-11eb-9a27-5b4b2ca5fa42.png)

1) why is the `EuiResizableButton` disabled when the panel has been collapsed?
2) if the toggle button is "on top / inside" the `EuiResizableButton` I think its location should be consistent whether the panel has been collapsed or not.

thompsongl commented:
> why is the EuiResizableButton disabled when the panel has been collapsed?

This was decided earlier in the thread: https://github.com/elastic/eui/pull/3978#issuecomment-689701728
If a panel is collapsed it doesn't make sense to allow it be resized.

> if the toggle button is "on top / inside" the EuiResizableButton I think its location should be consistent whether the panel has been collapsed or not.

I can make this change. There are mocks/screenshots earlier in the thread that show it both ways, so this works for me

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

thompsongl commented:
@andreadelrio This is ready for another design review. Definitely pay attention to the various toggle button visual states and any functionality that goes against your expectations

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

thompsongl commented:
Going to work on documentation before removing Draft status. Feel free to comment on design and functionality in the meantime

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

cchaos commented:
This is super slick. I think a lot of people will be happy with this. I know it's been no small effort, so thank you!!

When I'm playing with the current examples, they work excellently. What I start to think about now is use-cases, mainly the Lens application layout (which is going to be more common). The Lens layout colors the background of it's two side panels using a border to indicate where the resizable button would be.

<img width="50%" alt="Screen Shot 2020-10-29 at 11 27 32 AM" src="https://user-images.githubusercontent.com/549577/97595654-48cea800-19da-11eb-8f0f-6d9929ef277c.png">

Currently there's no way to achieve this because of the width of the resizable button. I hacked at the styles in the dev tools inspector to simulate and figure out what would be needed to achieve this. Here's a quick screencast: https://share.getcloudapp.com/2NujWrlY

Basically what is needed is:

- [ ] Resizable button to allow for a `0` width but in reality is at least 16px wide with -8x margins on the sides to overlap content
- [ ] Panel content needs to allow for padding to accomodate for inner content 
- [ ] When collapsed, the collapsible button should overlap the resizable button
- [ ] When collapsed, the panel should have 0 width (not a percentage)

---

As for the collapsible functionality specifically, I'd still love to try to find a way to allow the collapse button track under the mouse when in focus and feel that out. But maybe that can be a follow-up.

This is one of those components that we are creating  in EUI mainly for the sake of consistency of functionality and usability. I think it's important that we actually maintain consistency of look and behavior. Mainly, I'm trying to say that I think we should not allow consumers to change the collapsible button icon. By allowing customization we lose that recognition.

I think that if they don't want the collapsible button that is innate, they just need to **recreate** their own. This would then be more like how the Maps application would handle this by needing to control the position of this button anyway.

--- 

Other things I noticed:

- [ ] When collapsed, it's could be hard to find the uncollapse button. Can we not disable the resizable button so that the whole thing can be clicked to re-open?
- [ ] The rotated button for horizontal splits doesn't quite work because the lines are going the wrong direction. @andreadelrio could you create a new icon for this direction?
- [ ] The docs examples are wrapping everything in  an EuiText block which is causing weird margins for the list groups. Can you just move this to only wrap text?
- [ ] In the panel options example, I was expecting the "Add data" button to do something. 
- [ ] The last example with the external control didn't act as I was expecting. I know this logic is complicated but we should probably work on it more. Also, at one point I got the resizable buttons to not work (after lots of clicked the toggle buttons).

thompsongl commented:
Thanks for thorough review and clear requests, @cchaos 

This all makes sense to me, with just a couple questions below for clarification. I'll do more clean up work on the docs and probably change the "Add data" panel; just tried to make the example look more real, but don't want to make an app out of it.


> Panel content needs to allow for padding to accomodate for inner content

Do you mean that `EuiResizablePanel` should have a `paddingSize` prop like `EuiPanel`? Currently it's just an empty, unstyled `div` that consumers can style and add content (like an EuiPanel`)  as needed.

> When collapsed, it's could be hard to find the uncollapse button. Can we not disable the resizable button so that the whole thing can be clicked to re-open?

If we made these two items the standard behavior would it resolve your concerns?
* When collapsed, the collapsible button should overlap the resizable button
* When collapsed, the panel should have 0 width (not a percentage)

cchaos commented:
> Do you mean that EuiResizablePanel should have a paddingSize prop like EuiPanel? Currently it's just an empty, unstyled div that consumers can style and add content (like an EuiPanel`) as needed.

Yes because this is where the `overflow` is handled. Otherwise consumers would have to add an internal div, with padding, with forced stretch, with overflow. This is a lot to add for simply getting padding and/or a background color. In fact, if it doesn't already, the panel should also accept a custom classname.

<img width="897" alt="Screen Shot 2020-10-29 at 13 35 07 PM" src="https://user-images.githubusercontent.com/549577/97611115-204fa980-19ec-11eb-9b08-620d628d1f54.png">


Eventually we could even completely convert this to use EuiPanel and therefore inherit all it's properties. But that's blocked by #4194

> If we made these two items the standard behavior would it resolve your concerns?

I don't think those two bullets address the original ask which is more about having such a small target to re-open the panel. Especially one that **could** be hidden from view. My intent was to just say, never **disable** the resize button, but if the panel is collapsed make the only action on it be a single click to re-open.

thompsongl commented:
☑️ We can add some padding options, though I'd like to do so in a follow-up PR when the revamped `EuiPanel` is in a place to be considered for use from the start.

☑️ I'll explore changing the disabled state of the resizer to something interactive

thompsongl commented:
Removed icon customization in the latest commit.

> The last example with the external control didn't act as I was expecting

Found a bug where sometimes the calculated width was negative. Still solving.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

thompsongl commented:
Addressed most of the feedback thus far. The points related to zero-width resizable buttons is dependent on the point about configurable panel padding, which is currently blocked (in my view) by the EuiPanel rework. These two may need to be a follow-up PR.
Will wait for new icons before taking out of Draft.

andreadelrio commented:
> Addressed most of the feedback thus far. The points related to zero-width resizable buttons is dependent on the point about configurable panel padding, which is currently blocked (in my view) by the EuiPanel rework. These two may need to be a follow-up PR.
> Will wait for new icons before taking out of Draft.

I'll get to the icons asap.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

andreadelrio commented:
> * The rotated button for horizontal splits doesn't quite work because the lines are going the wrong direction. @andreadelrio could you create a new icon for this direction?

@cchaos Been exploring some options for these icons. Let me know what you think. 

![image](https://user-images.githubusercontent.com/4016496/98426492-eadf3780-204d-11eb-9cdf-abce9f56b48a.png)

With other icons for perspective
![image](https://user-images.githubusercontent.com/4016496/98426498-f3377280-204d-11eb-9fe2-8cccde11e2dd.png)

With grid
![image](https://user-images.githubusercontent.com/4016496/98426529-119d6e00-204e-11eb-8fa0-3044ac416b82.png)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

andreadelrio commented:
Created the PR with the icons for vertical collapse #4244

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

chandlerprall commented:
I don't think this is something other than "works as intended" but wanted to point it out in case. In the _Resizable panel options_ example, the middle panel can be resized so small that it hides behind the third panel.

![resizable_width](https://user-images.githubusercontent.com/313125/100795831-ea338a00-33dc-11eb-8d6c-3748a9baac88.gif)

thompsongl commented:
> "works as intended"

Exactly. We could update the min-width on that example to be a bit more realistic, I bet.

chandlerprall commented on src/components/resizable_container/resizable_collapse_button.tsx:
```diff
@@ -0,0 +1,98 @@
+    className
+  );
+
+  // Default to simiple grab icon in case there is no externalPosition specified
```
`externalPosition` is required and should always be specific?

thompsongl replied:
Question for @cchaos: Did you intend for `externalPosition` to be optional? I could see it both ways.

cchaos replied:
I think it was mainly a backup to the logic just in case, but in reality every instance does need an `externalPosition` to know which direction to indicate. This is a fallback in case a consumer isn't using TS and setting a default `externalPosition` isn't quite right because it's so context specific.

thompsongl replied:
👍 I'll make it optional

chandlerprall commented on src/components/resizable_container/resizable_container.tsx:
```diff
@@ -109,52 +112,155 @@ export const EuiResizableContainer: FunctionComponent<EuiResizableContainerProps
+  const onMouseDown = useCallback(
+    (event: EuiResizableButtonMouseEvent) => {
+      const currentTarget = event.currentTarget;
+      const prevPanelId = currentTarget.previousElementSibling!.id;
```
let's include a quick check for previousElementSibling and nextElementSibling to not be null (and return/noop if that's the case), since the inclusion/placement of the button is controlled by the consuming app

chandlerprall reviewed (changes requested):
Reviewed everything except the reducer's logic, will do that one in a second pass.

* The _Horizontal resizing with three panels_ example feels like it has been superseded by _Resizable panel options_
* Navigating to the collapse button via keyboard, there is no focus state on the button

thompsongl commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+
+  const [reducerState, dispatch] = useReducer(reducer, initialState, init);
+
+  // TODO: Handle side-effects like this?
```
Mostly left that in so you could see the other way it'd work. Unless you feel strongly that this is better, I'll remove it.

chandlerprall replied:
I think the useEffect is better, dispatching a new action from a reducer goes against Best Practices™️, and even with `runSideEffect` marked as async it still runs immediately - only now returns a promise that resolves to undefined the next event tick.

Another alternative structure would be to add the side effect call to the relevant actions after they make the first dispatch - that's where the reducer orchestration typically happens. In this case I think that setup is less clear because it isn't obvious what ties the three actions to that extra dispatch, whereas the `useEffect` explicitly lists the panels as a dependency.

thompsongl replied:
> I think the useEffect is better

To clarify, you prefer using:
1) _`runSideEffect`_ as-is (maybe not async)
2) `useEffect` and calling `onPanelWidthChange` directly from there and eliminating both the `EUI_RESIZABLE_ONCHANGE` dispatch and the `runSideEffect` wrapper

chandlerprall replied:
Option #2

thompsongl replied:
Remembering why I didn't do Option 2 originally: it results in infinite loops errors.
It results in `useEffect` calling `setState` without access to the dependencies of state (because the callback comes from a parent component). That's the reason for the indirectional dispatch method, and the reason I went with a more inline alternative.

chandlerprall replied:
👍 that makes sense; let's go with option 1 then (with removing `async`).

thompsongl commented on src/components/resizable_container/resizable_button.tsx:
```diff
@@ -53,58 +56,70 @@ export interface EuiResizableButtonProps
-  /**
-   * The size of the Resizer (the space between panels)
-   */
-  size?: EuiResizableButtonSize;
```
Yes, this is the sole breaking change. Marked the PR already with a label.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
-        );
-        const delta = x - state.currentResizerPos;
-        const containerSize = getContainerSize() - state.resizersSize;
+      case 'EUI_RESIZABLE_DRAG_MOVE': {
```
To be a little more defensive, this reducer could verify `state.isDragging` before making changes. If you hate the change, ignore this - I'm not sure how nit picky it is.

chandlerprall commented on src/components/resizable_container/resizable_container.tsx:
```diff
@@ -109,52 +112,158 @@ export const EuiResizableContainer: FunctionComponent<EuiResizableContainerProps
+
+  const onMouseMove = useCallback(
+    (event: React.MouseEvent | React.TouchEvent) => {
+      if (!reducerState.prevPanelId || !reducerState.nextPanelId) return;
```
Is testing for `isDragging` instead a better option? Or all three? Probably all 3 values, unless the state type is updated to reflect that the panel ids exist when isDragging is true, but that might be more work than worth. Your decision(s) 😀.

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+          state.containerSize
+        );
+
+        return withSideEffect({
```
This needs a minSize check like done in `EUI_RESIZABLE_DRAG_MOVE`

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+        const currentPanel = state.panels[currentPanelId];
+        const shouldCollapse = !currentPanel.isCollapsed;
+        const panelElement = document.getElementById(currentPanelId);
+        const prevResizer = panelElement!.previousElementSibling;
```
This is another place that relies on the consuming app to place the resizers in expected locations, can we add some quick logic here to hunt for to the button similar to `getSiblingPanel` ?

thompsongl replied:
This one is slightly different. We want to disabled to _immediate_ sibling resizer in this case, and there is an existence check for it later. Finding an adjecent-but-not-immediate-sibling resizer isn't important in this case because (in this very unlikely scenario), it wouldn't need to be disabled.

chandlerprall replied:
Trying to think through what a consumer could do to screw with this,

```
        <EuiResizablePanel initialSize={50} minSize="30%">
          <EuiText>
            <p>{text}</p>
            <a href="">Hello world</a>
          </EuiText>
        </EuiResizablePanel>

        <EuiResizableButton />

        <div className="funAnimationThing"/>

        <EuiResizablePanel initialSize={50} minSize="200px">
          <EuiText>
            <p>{text}</p>
          </EuiText>
        </EuiResizablePanel>
```

then the extra div would be selected as the resizer, right?

thompsongl replied:
Oh this would break everything regardless. The sizes would be off from the start. There can't be anything except EuiResizablePanel and EuiResizableButton as children. It's been that way since v1 of this component.

chandlerprall replied:
Fair enough!

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+          0
+        );
+
+        // A toggling sequence has occurred where a to-be-opened panel will
```
🚀

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+          // This situation can only occur when (n-1) panels are collapsed at once
+          // and the most recently collapsed panel gains significant width
+          // during the previously occurring collapse.
+          // That is (largley), external toggling where the default logic has
```
```suggestion
          // That is (largely), external toggling where the default logic has
```

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+          // That is (largley), external toggling where the default logic has
+          // been negated by the lack of panel mode distinction.
+          otherPanels = {
+            ...Object.values(state.panels).reduce(
```
Don't need to spread the resulting object into a new object :)

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+            (100 - collapsedPanelsSize) / (Object.keys(otherPanels).length + 1);
+
+          updatedPanels = {
+            ...Object.values(otherPanels).reduce(
```
Same spreading into an empty object here

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+        } else {
+          // A toggling sequence has occurred that is standard and predictable
+          updatedPanels = {
+            ...Object.values(otherPanels).reduce(
```
Here too

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+            ),
+          };
+        } else {
+          // A toggling sequence has occurred that is standard and predictable
```
> standard and predictable

if only everything could be that easy 😀

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+            },
+          },
+          resizers: {
+            ...Object.values(state.resizers).reduce(
```
Another spread into an empty object

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+        return {
+          ...state,
+          resizers: {
+            ...Object.values(state.resizers).reduce(
```
Another extra object creation

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+        return {
+          ...state,
+          resizers: {
+            ...Object.values(state.resizers).reduce(
```
🎏

chandlerprall commented on src/components/resizable_container/helpers.ts:
```diff
@@ -17,230 +17,607 @@
+        onPanelWidthChange!(sizesOnly(state.panels));
+        return state;
+      }
+      // TODO: Implement more generic version of
```
Something you want to do in this iteration?

thompsongl replied:
Nope. This can be a follow-up.

chandlerprall reviewed (changes requested):
The first example is no longer resizable by mouse 😕

thompsongl commented:
Still working on the unresolved comments.


> The Horizontal resizing with three panels example feels like it has been superseded by Resizable panel options

I agree; removed.

> Navigating to the collapse button via keyboard, there is no focus state on the button

Added some styles. Feedback welcome.

> The first example is no longer resizable by mouse

Fixed.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/

thompsongl commented on src-docs/src/views/resizable_container/resizable_panel_collapsible_external.js:
```diff
@@ -0,0 +1,125 @@
+];
+
+export default () => {
+  const collapseFn = useRef(() => {});
```
Nothing is immediately coming to mind for a refactor. The idea was to handle this similar to how `ref`s are handled.

chandlerprall reviewed (approved):
Changes LGTM! The +2,675 −620 modified lines of code over 30 files does not fully represent the work y'all put into this feature. This is huge, thank you!

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -7,48 +8,151 @@ import { GuideSectionTypes } from '../../components';
-const ResizableContainerThreePanelsHtml = renderToHtml(
-  ResizableContainerThreePanels
-);
+const basicSnippet = `<EuiResizableContainer style={{ height: '400px' }}>
```
```suggestion
const basicSnippet = `<EuiResizableContainer>
```

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -7,48 +8,151 @@ import { GuideSectionTypes } from '../../components';
+</EuiResizablePanel>`;
+
+const ResizableContainerVerticalHtml = renderToHtml(ResizableContainerVertical);
+const verticalSnippet = `<EuiResizableContainer direction="vertical" style={{ height: '400px' }}>
```
```suggestion
const verticalSnippet = `<EuiResizableContainer direction="vertical">
```

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -7,48 +8,151 @@ import { GuideSectionTypes } from '../../components';
 );
-const ResizableResizerSizeHtml = renderToHtml(ResizableResizerSize);
+const ResizablePanelCollapsibleHtml = renderToHtml(ResizablePanelCollapsible);
+const collapsibleSnippet = `<EuiResizableContainer style={{ height: '400px' }}>
```
```suggestion
const collapsibleSnippet = `<EuiResizableContainer>
```

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -7,48 +8,151 @@ import { GuideSectionTypes } from '../../components';
+  {(EuiResizablePanel, EuiResizableButton) => (
+    <>
+      <EuiResizablePanel initialSize={50} minSize="200px">
+        <EuiText>
+          <p>{text}</p>
+        </EuiText>
```
Let's also simplify all these snippets by removing the EuiText portion in favor of a comment so that they don't accidentally put non-text components inside or EuiText.
```suggestion
        <!-- Left panel content -->
```

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -7,48 +8,151 @@ import { GuideSectionTypes } from '../../components';
+      <EuiResizableButton />
+
+      <EuiResizablePanel initialSize={50} minSize="200px">
+        <EuiText>
+          <p>{text}</p>
+        </EuiText>
```
```suggestion
        <!-- Right panel content -->
```

cchaos replied:
And so on...

cchaos commented on src-docs/src/views/resizable_container/resizable_container_example.js:
```diff
@@ -63,22 +167,21 @@ export const ResizableContainerExample = {
   intro: (
     <Fragment>
       <EuiCallOut title="Consuming" color="warning">
```
Let's remove this intro from the callout. I don't think there's anything extra special or warning-like in this paragraph.

cchaos reviewed (approved):
Functionality all looks great. I'm sure we might find some more strange situations but, we won't know til we release 😄 

Just had a few more suggestions for the docs.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3978/
